Modules/_sre/sre.c

```diff
@@ -2715,6 +2824,107 @@ pattern_scanner(_sremodulestate *module_state,
     return (PyObject*) scanner;
 }
 
+/* -------------------------------------------------------------------- */
+/* template methods */
+
+static int
+template_traverse(TemplateObject *self, visitproc visit, void *arg)
+{
+    Py_VISIT(Py_TYPE(self));
+    Py_VISIT(self->literal);
+    for (Py_ssize_t i = 0, n = Py_SIZE(self); i < n; i++) {
+        Py_VISIT(self->items[i].literal);
+    }
+    return 0;
+}
+
+static int
+template_clear(TemplateObject *self)
+{
+    Py_CLEAR(self->literal);
+    for (Py_ssize_t i = 0, n = Py_SIZE(self); i < n; i++) {
+        Py_CLEAR(self->items[i].literal);
+    }
+    return 0;
+}
+
+static void
+template_dealloc(TemplateObject *self)
+{
+    PyTypeObject *tp = Py_TYPE(self);
+
+    PyObject_GC_UnTrack(self);
+    (void)template_clear(self);
+    tp->tp_free(self);
+    Py_DECREF(tp);
+}
+
+static PyObject *
+expand_template(TemplateObject *self, MatchObject *match)
+{
+    if (Py_SIZE(self) == 0) {
+        Py_INCREF(self->literal);
+        return self->literal;
+    }
+
+    PyObject *buffer[10];
+    PyObject **out = buffer;
+    Py_ssize_t count = 0;
+    PyObject *list = NULL;
+    if (self->chunks > 10 || !PyUnicode_Check(self->literal)) {
+        list = PyList_New(self->chunks);
+        if (!list) {
+            return NULL;
+        }
+        out = &PyList_GET_ITEM(list, 0);
+    }
+
+    Py_INCREF(self->literal);
+    out[count++] = self->literal;
+    for (Py_ssize_t i = 0; i < Py_SIZE(self); i++) {
+        Py_ssize_t index = self->items[i].index;
+        if (index >= match->groups) {
+            PyErr_SetString(PyExc_IndexError, "no such group");
+            Py_DECREF(list);
```

Comment:
`Py_XDECREF(list);`

Reply:
Oh, not only this. We need to clean up the buffer.

The support of the on-stack buffer was added recently, so I missed some corner cases.